stdlib/src/utils/span.mojo

```diff
@@ -256,3 +258,64 @@ struct Span[
         """
 
         return self._data
+
+    @always_inline
+    fn __bool__(self) -> Bool:
+        """Check if a span is non-empty.
+
+        Returns:
+           True if a span is non-empty, False otherwise.
+        """
+        return len(self) > 0
+
+    fn __eq__[
+        T2: ComparableCollectionElement
+    ](
+        self: Reference[Span[T2, is_mutable, lifetime]], rhs: Span[T, _, _]
+    ) -> Bool:
+        """Verify if span is equal to another span.
+
+        Parameters:
+            T2: The type of the elements in the span. Must implement the
+              traits `EqualityComparable` and `CollectionElement`.
+
+        Args:
+            rhs: The span to compare against.
+
+        Returns:
+            True if the spans are equal in length and contain the same elements, False otherwise.
+        """
+        constrained[_type_is_eq[T, T2](), "T must be equal to T2"]()
+        # both empty
+        if not self[] and not rhs:
+            return True
+        if len(self[]) != len(rhs):
+            return False
+        # same pointer and length, so equal
+        if self[].unsafe_ptr().bitcast[T]() == rhs.unsafe_ptr():
+            return True
+        for i in range(len(self[])):
+            if self[][i][] != rebind[T2](rhs[i][]):
+                return False
+        return True
+
+    @always_inline
+    fn __ne__[
+        T2: ComparableCollectionElement
+    ](
+        self: Reference[Span[T2, is_mutable, lifetime]], rhs: Span[T, _, _]
+    ) -> Bool:
+        """Verify if span is not equal to another span.
+
+        Parameters:
+            T2: The type of the elements in the span. Must implement the
+                traits `EqualityComparable` and `CollectionElement`.
+
+        Args:
+            rhs: The span to compare against.
+
+        Returns:
+            True if the spans are not equal in length or contents, False otherwise.
+        """
+        constrained[_type_is_eq[T, T2](), "T must be equal to T2"]()
+        return not self[] == (rhs)
```

Comment:
This is crashing the compiler in ParameterInference internally.  Can we try to add some more tests and/or look into a minimal repro to file a compiler bug?

Reply:
```cd /__w/modular/modular/.derived/build-release && /__w/modular/modular/.derived/build-release/bin/mojo package /__w/modular/modular/Kernels/mojo/stdlib -o Kernels/mojo/stdlib/stdlib.mojopkg
mojo: /__w/modular/modular/KGEN/lib/MojoParser/ParameterInference.cpp:744: Attribute getPartiallySpecializedSignature(ArrayRef<TypedAttr>, ParserParamEvaluator &, M::KGEN::LIT::LITSignatureType)::Substitutor::tryReplace(Attribute, size_t): Assertion `result.getType() == evaluator.getReboundType(ref.getType()) && "Parameter type mismatch"' failed.
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.	Program arguments: /__w/modular/modular/.derived/build-release/bin/mojo package /__w/modular/modular/Kernels/mojo/stdlib -o Kernels/mojo/stdlib/stdlib.mojopkg
1.	Crash resolving decl body at loc("/__w/modular/modular/Kernels/mojo/stdlib/utils/span.mojo":303:8)
    >>     fn __ne__[
              ^......
    >>         T2: ComparableCollectionElement
    >>     ](
    >>         self: Reference[Span[T2, is_mutable, lifetime]], rhs: Span[T, _, _]
    >>     ) -> Bool:
2.	Crash parsing statement at loc("/__w/modular/modular/Kernels/mojo/stdlib/utils/span.mojo":321:9)
    >>         return not self[] == (rhs)
```

after a nice multiple hundred frame stack trace. @laszlokindrat or @Mogball ideas here/suggestions welcome for chasing a minimal repro for parameter inference rebound type bug.

Reply:
The manual technique is just semi randomly deleting code from all over the place, including from the stdlib

Reply:
@lsh want to which to using `ref` arguments to see if the crash still happens?